grails-app/views/stockMovement/show.gsp

```diff
@@ -75,78 +77,119 @@
             <g:set var="hasBeenPlaced" value="${stockMovement?.hasBeenShipped() || stockMovement?.hasBeenPartiallyReceived()}"/>
             <g:set var="isSameOrigin" value="${stockMovement?.origin?.id==session.warehouse.id}"/>
             <g:set var="isSameDestination" value="${stockMovement?.destination?.id==session.warehouse.id}"/>
-            <g:if test="${stockMovement?.order}">
-                <g:link controller="stockTransfer" action="edit" id="${stockMovement?.order?.id}" class="button">
-                    <img src="${resource(dir: 'images/icons/silk', file: 'pencil.png')}" />&nbsp;
-                    <warehouse:message code="default.button.edit.label" />
-                </g:link>
-            </g:if>
-            <g:else>
-                <g:link controller="stockMovement" action="edit" id="${stockMovement?.id}" class="button">
-                    <img src="${resource(dir: 'images/icons/silk', file: 'pencil.png')}" />&nbsp;
-                    <warehouse:message code="default.button.edit.label" />
-                </g:link>
-            </g:else>
-            <g:link controller="partialReceiving" action="create" id="${stockMovement?.shipment?.id}" class="button">
-                <img src="${resource(dir: 'images/icons/', file: 'handtruck.png')}" />&nbsp;
-                <warehouse:message code="default.button.receive.label" />
-            </g:link>
-            <g:isUserAdmin>
-                <g:if test="${stockMovement?.hasBeenReceived() || stockMovement?.hasBeenPartiallyReceived()}">
-                    <g:link controller="partialReceiving" action="rollbackLastReceipt" id="${stockMovement?.shipment?.id}" class="button">
-                        <img src="${resource(dir: 'images/icons/silk', file: 'arrow_rotate_anticlockwise.png')}" />&nbsp;
-                        <warehouse:message code="stockMovement.rollbackLastReceipt.label" />
+            <g:set var="userHasRequestApproverRole" value="${false}"/>
+            <g:userHasRoles location="${stockMovement?.origin?.id}" roles="${[RoleType.ROLE_REQUISITION_APPROVER]}">
```

Comment:
I know @jmiranda is not a big fan of that, but I don't know if I see any other possible solution at the moment
But I like the fact, that you've actually checked it once, and assigned it to a variable, so we wouldn't have to call this `<g:userHasRoles>` in many places, like we do for `<g:isSuperuser>` in many places

Reply:
Yeah, I hate this paradigm, but I think I'm the one who made it popular so I don't blame @drodzewicz.

What is interesting is that the <g:set /> actually gets invoked with that. I guess that makes sense, but I assumed g:if did something special. I guess it's just that GSP is rendered/invoked recursively so whether the next tag is rendering or setting doesn't really matter ... it just gets invoked. But it would be great  if someone investigated the Grails tab library where "g:if" is defined to see what they do in this case.

And as I said above, I would prefer `ifUserHasRoles` over `userHasRoles`.

Reply:
It probably needs to be done as a tech debt refactoring but we probably should have thought about converging our permission/supports taglibs with something closer to Spring Security (like using `authorize` with logic expressions).